Wikipedia/Code/SchemeHandler/SchemeHandler.swift

```diff
@@ -107,37 +112,47 @@ extension SchemeHandler: WKURLSchemeHandler {
 
         switch baseComponent {
         case FileHandler.basePath:
-            fileHandler.handle(pathComponents: pathComponents, requestURL: requestURL, completion: localCompletionBlock)
+            fileHandler.handle(pathComponents: pathComponents, requestURL: originalRequestURL, completion: localCompletionBlock)
             
         default:
             
-            guard let defaultURL = defaultHandler.urlForPathComponents(pathComponents, requestURL: requestURL) else {
-                urlSchemeTask.didFailWithError(SchemeHandlerError.invalidParameters)
-                removeSchemeTask(urlSchemeTask: urlSchemeTask)
-                return
+            guard let requestURL = defaultHandler.urlForPathComponents(pathComponents, requestURL: originalRequestURL),
+                let request = cacheController?.newCachePolicyRequest(from: originalRequest as NSURLRequest, newURL: requestURL) else {
+                    urlSchemeTask.didFailWithError(SchemeHandlerError.invalidParameters)
+                    removeSchemeTask(urlSchemeTask: urlSchemeTask)
+                    return
             }
+            
             // IMPORTANT: Ensure the urlSchemeTask is not strongly captured by this block operation
             // Otherwise it will sometimes be deallocated on a non-main thread, causing a crash https://phabricator.wikimedia.org/T224113
             let op = BlockOperation { [weak urlSchemeTask] in
-                if let cachedResponse = self.articleCacheController?.recentCachedURLResponse(for: defaultURL) {
+                
+                //forceCache will be true for ArticleViewControllers when coming from Navigation State Controller. In this case stale data is fine.
+                if self.forceCache,
```

Comment:
Is this alright without `forceCache` here and on the `ArticleViewController`? I think we should be always OK with the cached data unless I'm missing something

Reply:
@joewalsh should be, the default functionality (when this flag is false) will attempt to pull from cache first if the response header rules indicate that. The mobile-html response header for `cache-control `returns `max-age=0` and `must-revalidate`, so it first validates it with the etag and if it sees 304 then pulls from cache (URLCache then fallback to PersistentCache). This `forceCache` flag bypasses this step and goes straight to pulling from cache if we know there's anything there. So it won't bother even validating to check for a 304 and will see stale data if edits have been made since then. For now this is only used when we restore their last read article upon launch.

I realize this flag sortof duplicates a `NSURLRequest.CachePolicy.returnCacheDataElseLoad` policy, but if I went with that policy I think I would have had to subclass `URLCache` and hit this PersistentCache system within it for the PersistentCache fallback functionality (this is how the lite project did it in with `PermanentlyPersistableURLCache`). Instead this flows through `CacheController.cachedURLResponse(for request: URLRequest)` which decides between URLCache or PersistentCache. Just felt a little more targeted to me than having every call that goes through URLCache (calls that have nothing to do with saved articles) potentially touch PersistentCache under the hood.

Reply:
@tonisevener makes sense, thanks for the explanation! I see now that it's helpful to have this flag to avoid the pre-request check and agree it's more targeted this way